drivers/platform/maxim/max32670/maxim_gpio_irq.c

```diff
@@ -0,0 +1,381 @@
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+
+/******************************************************************************/
+/***************************** Include Files **********************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <stdint.h>
+#include <errno.h>
+
+#include "no_os_error.h"
+#include "no_os_util.h"
+#include "no_os_list.h"
+#include "no_os_irq.h"
+#include "no_os_gpio.h"
+
+#include "maxim_gpio_irq.h"
+#include "maxim_irq.h"
+
+/******************************************************************************/
+/*************************** Types Declarations *******************************/
+/******************************************************************************/
+
+static struct no_os_list_desc *actions;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+/**
+ * @brief GPIO callback function that sets the event and further calls
+ * the user registered callback
+ * @param cbdata - Used to pass an action reference
+ */
+static void gpio_irq_callback(void *cbdata)
+{
+	int ret;
+	struct irq_action *action;
+	struct irq_action key;
+
+	key.irq_id = ((struct irq_action *)cbdata)->irq_id;
+	ret = no_os_list_read_find(actions, (void **)&action, &key);
+	if (ret)
+		return;
+
+	if (action->callback)
+		action->callback(action->ctx);
+}
+
+void GPIO0_IRQHandler()
+{
+	MXC_GPIO_Handler(0);
+}
+
+#ifdef MXC_GPIO1
+void GPIO1_IRQHandler()
+{
+	MXC_GPIO_Handler(1);
+}
+#endif
+
+#ifdef MXC_GPIO2
+void GPIO2_IRQHandler()
+{
+	MXC_GPIO_Handler(2);
+}
+#endif
+
+/**
+ * @brief Initialize the GPIO interrupt controller
+ * @param desc - Pointer where the configured instance is stored
+ * @param param - Configuration information for the instance
+ * @return 0 in case of success, errno error codes otherwise.
+ */
+static int max_gpio_irq_ctrl_init(struct no_os_irq_ctrl_desc **desc,
+				  const struct no_os_irq_init_param *param)
+{
+	int ret;
+	struct no_os_irq_ctrl_desc *descriptor;
+
+	if (!param)
+		return -EINVAL;
+
+	descriptor = calloc(1, sizeof(*descriptor));
+	if (!descriptor)
+		return -ENOMEM;
+
+	descriptor->irq_ctrl_id = param->irq_ctrl_id;
+	descriptor->platform_ops = &max_gpio_irq_ops;
+	descriptor->extra = param->extra;
+
+	ret = no_os_list_init(&actions, NO_OS_LIST_PRIORITY_LIST, irq_action_cmp);
+	if (ret)
+		goto error;
+
+	*desc = descriptor;
+
+	return 0;
+
+error:
+	no_os_list_remove(actions);
```

Comment:
this is not needed. the `free` happens already in `the no_os_list_init` if errors occur.

Reply:
Make sense, will remove this check.